### ISSUE TITLE
Propagating wildcard types in non-homomorphic mapped types breaks styled-system types

### ISSUE THREAD

**Issue by sandersn** [microsoft/TypeScript#41790]
#41622 breaks `@types/styled-system`. Here's a simplified example:

```ts
export type TV<T, K extends keyof T> = T[K] extends Record<infer E, any> ? E : never;
```

**Expected behavior:**
I can't tell what the expected behaviour is, but previously there was no error on `infer E`.

**Actual behavior:**
After #41622, an error on infer E: "Type 'E' does not satisfy the constraint 'string | symbol | number'."

**Comment by sandersn** [comment]
The original code is

```ts
export type ObjectOrArray<T, K extends keyof any = keyof any> = T[] | Record<K, T | Record<K, T> | T[]>;
export type ThemeValue<K extends keyof ThemeType, ThemeType, TVal = any> =
    ThemeType[K] extends TVal[] ? number :
    ThemeType[K] extends Record<infer E, TVal> ? E :
    ThemeType[K] extends ObjectOrArray<infer F> ? F : never;
```

**Comment by ahejlsberg** [comment]
This is a bug. We're being a little too aggressive in the wildcard type check and that causes a circularity. Also, looks like we aren't handling this scenario correctly:

```ts
type Foo<T> = T extends { [P in infer E]: any } ? E : never;
```

We end up reporting `Type 'E' is not assignable to type 'string | number | symbol'` when instead we should implicitly infer a constraint for `E`.

I'll put up a PR that fixes this.
